File: maelstromLibrary/src/main/java/ftc/library/MaelstromControl/MaelstromPID/PIDController.java
package ftc.library.MaelstromControl.MaelstromPID;

import ftc.library.MaelstromSensors.MaelstromTimer;
import ftc.library.MaelstromUtils.LibConstants;

/*class for PID controller*/
public class PIDController implements LibConstants {
    private double i=0;
    private double d;
    private double error;
    private double power;
    private double KP;
    private double KI;
    private double KD;
    private double previousError = 0;
    private double maxI;
    private double previousTime = 0;
    private MaelstromTimer loopTimer = new MaelstromTimer();


    public PIDController(double KP, double KI, double KD, double maxI){
        this.KP = KP;
        this.KI = KI;
        this.KD = KD;
        this.maxI = maxI;
    }

    public PIDController(double KP, double KI, double KD){
        this.KP = KP;
        this.KI = KI;
        this.KD = KD;
        maxI = 0;
    }
    public PIDController(double KP, double KI){
        this.KP = KP;
        this.KI = KI;
        maxI = 0;
    }


    public double power(double target, double currentLoc){
        double error = target - currentLoc;
        this.error = error;
        double deltaTime = (System.nanoTime() - previousTime)/NANOSECS_PER_MIN/*loopTimer.milliSecs()*/;
        //loopTimer.reset();
        i += Math.abs(currentLoc) > Math.abs(target) * 0.7 ? error*deltaTime : 0;
        if(maxI != 0) i = Math.min(maxI, Math.max(-maxI, i));
        d = (error - previousError)/deltaTime;
        previousTime = System.nanoTime();
        previousError = error;
        power = (KP*error) + (KI*i) + (KD*d);
        return power;
    }

    public double[] returnVal(){
        double PID[] = {(KP * error), (KI * i), (KD * d), power};
        return PID;
    }

    public double getP(){ return KP*error; }
    public double getI(){ return KI*i; }
    public double getD(){ return KD*d; }
    public double getError(){return Math.abs(error);}

    public double getIntegral() {
        return i;
    }

    public double getKP() {
        return KP;
    }
    public double getKI() {
        return KI;
    }
    public double getKD() {
        return KD;
    }

    public void setKP(double KP){this.KP = KP;}
    public void setKI(double KI){this.KI = KI;}
    public void setKD(double KD){this.KD = KD;}

    public void setPID(double kp, double ki, double kd){
        setKP(kp);
        setKI(ki);
        setKD(kd);
    }

    public void reset(){
        i = 0;
        previousError = 0;
    }


}

File: maelstromLibrary/src/main/java/ftc/library/MaelControl/PID/PIDController.kt
package ftc.library.MaelControl.PID

import ftc.library.MaelSensors.MaelTimer
import ftc.library.MaelUtils.LibConstants

/*class for PID controller*/
class PIDController : LibConstants {
    var integral: Double = 0.0
        private set
    private var d: Double = 0.0
    private var error: Double = 0.0
    private var power: Double = 0.0
    var kp: Double = 0.0
    var ki: Double = 0.00
    var kd: Double = 0.0
    private var previousError: Double = 0.0
    private var maxI: Double = 0.0
    private var previousTime = 0.0
    private val loopTimer = MaelTimer()

    val P: Double
        get() = kp * error
    val I: Double
        get() = ki * integral
    val D: Double
        get() = kd*d

    constructor(KP: Double, KI: Double, KD: Double, maxI: Double) {
        this.kp = KP
        this.ki = KI
        this.kd = KD
        this.maxI = maxI
    }

    constructor(KP: Double, KI: Double, KD: Double) {
        this.kp = KP
        this.ki = KI
        this.kd = KD
        maxI = 0.0
    }

    constructor(KP: Double, KI: Double) {
        this.kp = KP
        this.ki = KI
        maxI = 0.0
    }

    fun power(target: Double, currentLoc: Double): Double {
        val error = target - currentLoc
        this.error = error
        val deltaTime = (System.nanoTime() - previousTime) / LibConstants.NANOSECS_PER_MIN/*loopTimer.milliSecs()*/
        //loopTimer.reset();
        if (Math.abs(currentLoc) > Math.abs(target) * 0.7){
        integral +=  error * deltaTime}
        else integral = 0.0
        if (maxI != 0.0) integral = Math.min(maxI, Math.max(-maxI, integral))
        d = (error - previousError) / deltaTime
        previousTime = System.nanoTime().toDouble()
        previousError = error
        //power = kp * error + ki * integral + kd * d
        power = P + I + D
        return power
    }

    fun returnVal(): DoubleArray {
        return doubleArrayOf(kp * error, ki * integral, kd * d, power)
    }

    fun getError() = Math.abs(error)

    fun setPID(kp: Double, ki: Double, kd: Double) {
        this.kp = kp
        this.ki = ki
        this.kd = kd
    }

    fun reset() {
        integral = 0.0
        previousError = 0.0
    }


}